### PR TITLE
V2.0 css optimizer client

### PR DESCRIPTION
**Pull request by noahadams**
This is some client code for the upcoming css features of jazzcat, review would be appreciated.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -43,6 +43,26 @@ Utils.clone = function(obj) {
     return target;
 };
 
+// Some url helpers
+/**
+ * Takes a url, relative ro absolute, and absolutizes it relative to the current
```
Comment (jansepar) [inline src/utils.js]: to*

File: src/utils.js [hunk 2]
```diff
@@ -43,6 +43,26 @@ Utils.clone = function(obj) {
     return target;
 };
 
+// Some url helpers
+/**
+ * Takes a url, relative ro absolute, and absolutizes it relative to the current 
+ * document's location, with the assistance of an a element.
+ */
+var _absolutifyAnchor = document.createElement("a");
+Utils.absolutify = function(url) {
```
Comment (jansepar) [inline src/utils.js:52]: Useful util method! We use this in `resizeImages.js` as well (and maybe `capture.js`?), could you update those libraries to use this method as well?

Comment (jansepar) [inline src/utils.js:52]: Oh, and probably `jazzcat.js` :)

File: src/utils.js [hunk 3]
```diff
@@ -43,6 +43,26 @@ Utils.clone = function(obj) {
...
+    _absolutifyAnchor.href = url;
+    return _absolutifyAnchor.href;
+};
+
+/**
+ * Takes an absolute url, returns true if it is an http/s url, false otherwise 
+ * (e.g. mailto:, gopher://, etc.)
+ */
+var _httpUrlRE = /^https?/;
+Utils.httpUrl = function(url) {
```
Comment (jansepar) [inline src/utils.js:62]: Same comment as the one above: please use this in other libraries to DRY things up.

File: src/cssOptimize.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/**
+ * cssOptimize - Client code to a css optimization service
+ */
+
+define(["utils"], function(Utils) {
+
+var CssOptimize = window.cssOptimize = {};
+
+var absolutify = document.createElement("a");
```
Comment (jansepar) [inline src/cssOptimize.js]: Pew pew pew

File: src/cssOptimize.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * Takes an original, absolute url of a stylesheet, returns a url for that
+ * stylesheet going through the css service.
+ */
+
+CssOptimize.getCssURL = function(url, options) {
```
Comment (jansepar) [inline src/cssOptimize.js]: Wouldn't the convention here be `getCssUrl` ? We could just make it consistent with jazzcat and call it `getURL`

File: src/cssOptimize.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        url = Utils.absolutify(attributeVal);
+        if (Utils.httpUrl(url)) {
+            element.setAttribute(options.targetAttribute,
+                                 CssOptimize.getCssURL(url));
+        }
+    }
+};
+
+/**
+ * Takes an array-like object of `<link>` elements
```
Comment (jansepar) [inline src/cssOptimize.js:49]: We don't do any guarding against non-link elements. Should we throw in return if element.nodeName !== 'LINK' ? We also don't do that for `combineScripts` in jazzcat...

File: src/cssOptimize.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+};
+
+/**
+ * An 'error' event handler designed to be set using an "onerror" attribute that
+ * will set the target elements "href" attribute to the value of its 
+ * "data-orig-href" attribute, if one exists.
+ */
+var restoreOriginalHref = CssOptimize.restoreOriginalHref = function(event) {
```
Comment (jansepar) [inline src/cssOptimize.js:71]: We do the same thing for images. What do you think of making this a utility function to DRY things up and reduce LOC.

File: src/cssOptimize.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+/**
+ * Rewrite the href of a stylesheet referencing `<link>` element to go through 
+ * our service.
+ */
+CssOptimize._rewriteHref = function(element, options) {
+    var attributeVal = element.getAttribute(options.targetAttribute);
+    var url;
+    if(attributeVal) {
```
Comment (jansepar) [inline src/cssOptimize.js]: `if (attributeVal) {`

File: tests/cssOptimize.html [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            var l = document.createElement('link'),
+                origHref = 'bar';
+            l.setAttribute('data-orig-href', origHref);
+            l.setAttribute('href', 'foo');
+
+            cssOptimize.restoreOriginalHref({target: l});
+
+            strictEqual(l.getAttribute('href'), origHref);
+        });
+    });
```
Comment (jansepar) [inline tests/cssOptimize.html:85]: Could you add a test for `_rewriteHref` and `optimize`?

File: src/cssOptimize.js (hunk 6 above)
Comment (noahadams) [inline src/cssOptimize.js:49]: look down 8 lines

Comment (jansepar) [inline src/cssOptimize.js:49]: Hah. Yeah, not sure how I missed that :)

File: tests/cssOptimize.html (hunk 9 above)
Comment (noahadams) [inline tests/cssOptimize.html:85]: optimize is the first test, adding one for rewriteHref

Comment (jansepar) [inline tests/cssOptimize.html:85]: Ah, I missed that. Could you check that first tests description to "cssOptimize - optimize" to keep things consistent?

**Comment by jansepar** [comment]
Could you throw up an example here that I can use to test this bad boy out? Also since most of our APIs have 1 or more examples, it would be great to have one for cssOptimize as well.

File: src/cssOptimize.js (hunk 7 above)
Comment (noahadams) [inline src/cssOptimize.js:71]: we'd need to parameterize the attributes then, and invoke with the correct ones.

Comment (jansepar) [inline src/cssOptimize.js:71]: Yep. Change it to something like `restoreOriginalAttr` or something.

**Comment by noahadams** [comment]
@jansepar I've included an example, it doesn't show much visually but you can inspect to see where the stylesheet is coming from (and for now, notice that the error handler works, and if you run it locally against jazzcat-next, that localhost blacklisting works.

**Comment by jansepar** [comment]
Looking good! Giving this a conditional :+1: pending a merge and deploy of the CSS Optimization code in Jazzcat.

**Comment by noahadams** [comment]
Thanks!

**Pull request merged**